### PR TITLE
Add support for the Preonic rev3 board to ZMK

### PR DESCRIPTION
**Pull request by shurikai**
This pull request is to add support for the Preonic rev 3. The preonic is a board with built in MCU, and not a shield. The included default keymap is based on the one found on the QMK configurator, but will not be fully functional until shifted keycodes are available. 

To build: `west build -b preonic_rev3`

Note that there is one file touched below (kscan_gpio_matrix.c) that was strictly a change to line formatting in the C file. Nothing changed functionally.

### CODE REVIEW

File: app/boards/arm/preonic/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: MIT
```
Comment (petejohanson) [inline app/boards/arm/preonic/CMakeLists.txt:1]: Please include the copyright header here as well for the ZMK Contributors.

File: app/boards/arm/preonic/board.cmake [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: MIT
```
Comment (petejohanson) [inline app/boards/arm/preonic/board.cmake:1]: Ditto.

File: app/boards/arm/preonic/preonic_rev3.dts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+/ {
+	model = "Preonic PCD, rev3";
+	compatible = "preonic,rev3", "st,stm32f303";
+
+	chosen {
+		zephyr,sram = &sram0;
+		zephyr,flash = &flash0;
+		zmk,kscan = &kscan0;
+        zmk,matrix_transform = &layout_grid_transform;
```
Comment (petejohanson) [inline app/boards/arm/preonic/preonic_rev3.dts:19]: Some indenting inconsistency here.

File: app/boards/arm/preonic/preonic_rev3.dts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+RC(2,0) RC(2,1) RC(2,2) RC(2,3) RC(2,4) RC(2,5) RC(6,0) RC(6,1) RC(6,2) RC(6,3) RC(6,4) RC(6,5) 
+RC(3,0) RC(3,1) RC(3,2) RC(3,3) RC(3,4) RC(3,5) RC(7,0) RC(7,1) RC(7,2) RC(7,3) RC(7,4) RC(7,5)
+RC(8,0) RC(8,1) RC(8,2) RC(9,3)       RC(9,5)      RC(9,1)      RC(9,2) RC(8,3) RC(8,4) RC(8,5)
+		>;
+    };
+
+	kscan0: kscan_0 {
+		compatible = "zmk,kscan-gpio-matrix";
+		label = "KSCAN";
+        diode-direction = "col2row";
```
Comment (petejohanson) [inline app/boards/arm/preonic/preonic_rev3.dts:64]: Ditto.

File: app/boards/arm/preonic/preonic_rev3.keymap [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+#include <behaviors.dtsi>
+#include <dt-bindings/zmk/keys.h>
+
+#define DEFAULT 0
+#define LOWER   1
+#define RAISE   2
+#define ADJ     3
+
+/ {
+  chosen { zmk,matrix_transform = &layout_grid_transform; };
```
Comment (petejohanson) [inline app/boards/arm/preonic/preonic_rev3.keymap:15]: Ditto, some indent issues.

File: app/boards/arm/preonic/preonic_rev3.keymap [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+#define LOWER   1
+#define RAISE   2
+#define ADJ     3
+
+/ {
+  chosen { zmk,matrix_transform = &layout_grid_transform; }; 
+	keymap {
+		compatible = "zmk,keymap";
+
+		default_layer {
```
Comment (petejohanson) [inline app/boards/arm/preonic/preonic_rev3.keymap:19]: We've been trying to add the `&bt` bindings to stock keymaps *somewhere* so folks can use them. Can we find a spot in some layer?

File: app/drivers/zephyr/kscan_gpio_matrix.c [hunk 7]
```diff
@@ -51,6 +51,7 @@ static int kscan_gpio_config_interrupts(struct device **devices,
 }
 #endif
 
+
```
Comment (petejohanson) [inline app/drivers/zephyr/kscan_gpio_matrix.c:54]: Revert?

**Review by petejohanson — CHANGES_REQUESTED** [review]
A couple minor comments, and I believe the clang-format CI job may be failing due to issue with when you forked `main`. Can you rebase or merge up `main` to try to get that green again? Thanks!

File: app/boards/arm/preonic/CMakeLists.txt (hunk 1 above)
Comment (shurikai) [reply app/boards/arm/preonic/CMakeLists.txt:1]: Done.

File: app/boards/arm/preonic/board.cmake (hunk 2 above)
Comment (shurikai) [reply app/boards/arm/preonic/board.cmake:1]: Updated.

File: app/boards/arm/preonic/preonic_rev3.dts (hunk 3 above)
Comment (shurikai) [reply app/boards/arm/preonic/preonic_rev3.dts:19]: Corrected all the formatting issues in this file.

File: app/boards/arm/preonic/preonic_rev3.keymap (hunk 5 above)
Comment (shurikai) [reply app/boards/arm/preonic/preonic_rev3.keymap:15]: Corrected all the formatting issues in this file.

File: app/boards/arm/preonic/preonic_rev3.dts (hunk 4 above)
Comment (shurikai) [reply app/boards/arm/preonic/preonic_rev3.dts:64]: Corrected all the formatting issues in this file.

File: app/drivers/zephyr/kscan_gpio_matrix.c (hunk 7 above)
Comment (shurikai) [reply app/drivers/zephyr/kscan_gpio_matrix.c:54]: I'm not sure I follow the question. Can you elaborate?

File: app/boards/arm/preonic/preonic_rev3.keymap (hunk 6 above)
Comment (shurikai) [reply app/boards/arm/preonic/preonic_rev3.keymap:19]: No problem. I can add those.

Comment (shurikai) [reply app/boards/arm/preonic/preonic_rev3.keymap]: Done.

File: app/boards/arm/preonic/preonic_rev3.keymap [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        adjust {
+// -----------------------------------------------------------------------------------------
+// |   F1  |  F2  |  F3  |  F4  |  F5  |  F6  |  F7   |   F8  |  F9  |  F10 |  F11  |  F12  |
+// |       |RESET |      |      |      |      |       |       |      |      |       |  DEL  |
+// |       |      |      |      |      |      |       |       |      |      |       |       |
+// |       |      |      |      |      |      |       |       |      |      |       |       |
+// |       |      |      |      |      |      |       |       |      |      |       |       |
+            bindings = <
+    &kp F1       &kp F2       &kp F3       &kp F4       &kp F5       &kp F6  &kp F7  &kp F8  &kp F9  &kp F10 &kp F11 &kp F12   
+    &bt BT_SEL 0 &bt BT_SEL 1 &bt BT_SEL 2 &bt BT_SEL 3 &bt BT_SEL 4 &trans  &trans  &trans  &trans  &trans  &trans  &kp DEL
```
Comment (petejohanson) [inline app/boards/arm/preonic/preonic_rev3.keymap:74]: Thanks for adding this... At the same time... There's no way to *get* to this adjust layer (yet, I'm still brainstorming how best to handle this common pattern). Can we put this on `raise` instead?

**Review by petejohanson — CHANGES_REQUESTED** [review]
One last lingering keymap question, then we should get this in!

File: app/boards/arm/preonic/preonic_rev3.keymap (hunk 8 above)
Comment (petejohanson) [reply app/boards/arm/preonic/preonic_rev3.keymap:74]: @shurikai Did you see this comment? TIA.

**Comment by innovaker** [comment]
Is this PR dead?

**Comment by petejohanson** [comment]
@shurikai If still interested in getting this in, can you please see the one comment, and add a ZMK metadata file, so we can get this merged? 

Closing for now, but feel free to reopen.

**Pull request closed without merging**